File: src/time.ts
/**
 * Every time and date in this app is shown in India Standard Time (UTC+05:30),
 * on a 12-hour clock, regardless of what offset the API sends or what time zone
 * the machine is set to. Internally everything is absolute UTC milliseconds.
 */

export const MIN = 60_000;
export const IST_OFFSET = 330; // minutes east of UTC

const MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"] as const;
const DAYS = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"] as const;

const ISO =
  /^(\d{4})-(\d{2})-(\d{2})[T ](\d{2}):(\d{2})(?::(\d{2}))?(?:\.\d+)?(Z|[+-]\d{2}:?\d{2})?$/;

/** Parse an ISO timestamp to UTC ms. A missing offset is read as IST. */
export function parseISO(value: string | null | undefined): number | null {
  if (!value) return null;
  const m = ISO.exec(String(value).trim());
  if (!m) {
    const d = new Date(value);
    return Number.isNaN(d.getTime()) ? null : d.getTime();
  }
  const [, y, mo, d, h, mi, s, zone] = m;

  let offset = IST_OFFSET;
  if (zone) {
    if (zone === "Z") offset = 0;
    else {
      const z = /([+-])(\d{2}):?(\d{2})/.exec(zone);
      if (z) offset = (z[1] === "-" ? -1 : 1) * (Number(z[2]) * 60 + Number(z[3]));
    }
  }
  return (
    Date.UTC(Number(y), Number(mo) - 1, Number(d), Number(h), Number(mi), Number(s ?? 0)) -
    offset * MIN
  );
}

/** Shift into IST so the getUTC* accessors read as Indian wall-clock time. */
const ist = (t: number): Date => new Date(t + IST_OFFSET * MIN);
const p2 = (n: number): string => (n < 10 ? `0${n}` : String(n));
const h12 = (d: Date): number => d.getUTCHours() % 12 || 12;
const mer = (d: Date): string => (d.getUTCHours() >= 12 ? "PM" : "AM");

/** `09:47 AM` */
export function clockTime(t: number | null): string {
  if (t == null) return "--:--";
  const d = ist(t);
  return `${p2(h12(d))}:${p2(d.getUTCMinutes())} ${mer(d)}`;
}

/** `9:47 AM` — no leading zero, for headline use. */
export function clockShort(t: number | null): string {
  if (t == null) return "--:--";
  const d = ist(t);
  return `${h12(d)}:${p2(d.getUTCMinutes())} ${mer(d)}`;
}

/** `03:00:07 PM` */
export function clockWithSeconds(t: number): string {
  const d = ist(t);
  return `${p2(h12(d))}:${p2(d.getUTCMinutes())}:${p2(d.getUTCSeconds())} ${mer(d)}`;
}

/** `07/08/2026` — Indian day-first date format. */
export function dateDMY(t: number): string {
  const d = ist(t);
  return `${p2(d.getUTCDate())}/${p2(d.getUTCMonth() + 1)}/${d.getUTCFullYear()}`;
}

/** `Fri, 07 Aug 2026` */
export function dateFull(t: number): string {
  const d = ist(t);
  return `${DAYS[d.getUTCDay()]}, ${p2(d.getUTCDate())} ${MONTHS[d.getUTCMonth()]} ${d.getUTCFullYear()}`;
}

/** `FRIDAY` */
export function weekday(t: number): string {
  const full = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"] as const;
  return full[ist(t).getUTCDay()] ?? "";
}

/** `06 Aug 2026` */
export function dateLabel(t: number): string {
  const d = ist(t);
  return `${p2(d.getUTCDate())} ${MONTHS[d.getUTCMonth()]} ${d.getUTCFullYear()}`;
}

export function dayKey(t: number): string {
  const d = ist(t);
  return `${d.getUTCFullYear()}-${p2(d.getUTCMonth() + 1)}-${p2(d.getUTCDate())}`;
}

/** `2026-08` — the key a stored day is filed and filtered under. */
export function monthKey(t: number): string {
  const d = ist(t);
  return `${d.getUTCFullYear()}-${p2(d.getUTCMonth() + 1)}`;
}

/** `Aug 2026` */
export function monthLabel(t: number): string {
  const d = ist(t);
  return `${MONTHS[d.getUTCMonth()]} ${d.getUTCFullYear()}`;
}

/** `Aug 2026` from a stored `2026-08` key, for a month with no days left in it. */
export function monthLabelFromKey(key: string): string {
  const m = /^(\d{4})-(\d{2})$/.exec(key);
  const name = m ? MONTHS[Number(m[2]) - 1] : undefined;
  return m && name ? `${name} ${m[1]}` : key;
}

/** `07 Aug` — day and month only, for dense history rows. */
export function dayMonth(t: number): string {
  const d = ist(t);
  return `${p2(d.getUTCDate())} ${MONTHS[d.getUTCMonth()]}`;
}

/** Midnight IST of the day containing `t`, as UTC ms. */
export function startOfDay(t: number): number {
  const d = ist(t);
  return Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate()) - IST_OFFSET * MIN;
}

export const startOfToday = (): number => startOfDay(Date.now());

/** Midnight IST of a `YYYY-MM-DD` key — what an `<input type="date">` hands back. */
export function startOfDayKey(key: string): number | null {
  const m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(key.trim());
  if (!m) return null;
  const [, y, mo, d] = m;
  const t = Date.UTC(Number(y), Number(mo) - 1, Number(d)) - IST_OFFSET * MIN;
  return Number.isNaN(t) ? null : t;
}

/** `7:32` */
export function hm(ms: number): string {
  const neg = ms < 0;
  const total = Math.floor(Math.abs(ms) / MIN);
  return `${neg ? "-" : ""}${Math.floor(total / 60)}:${p2(total % 60)}`;
}

/** `7:32:09` */
export function hms(ms: number): string {
  const s = Math.floor(Math.abs(ms) / 1000);
  return `${Math.floor(s / 3600)}:${p2(Math.floor(s / 60) % 60)}:${p2(s % 60)}`;
}

/** Split for per-digit animation: `["7", "32", "09"]` */
export function hmsParts(ms: number): [string, string, string] {
  const s = Math.floor(Math.abs(ms) / 1000);
  return [String(Math.floor(s / 3600)), p2(Math.floor(s / 60) % 60), p2(s % 60)];
}

/** `8h 00m` — padded, for the goal headline. */
export function hoursMinutes(ms: number): string {
  const total = Math.round(Math.abs(ms) / MIN);
  return `${Math.floor(total / 60)}h ${p2(total % 60)}m`;
}

/** `1h 26m` */
export function human(ms: number): string {
  const total = Math.round(Math.abs(ms) / MIN);
  const h = Math.floor(total / 60);
  return `${h ? `${h}h ` : ""}${total % 60}m`;
}

/** Current IST wall-clock as `"14:05"`, ready for an `<input type="time">`. */
export function inputTimeNow(): string {
  const d = ist(Date.now());
  return `${p2(d.getUTCHours())}:${p2(d.getUTCMinutes())}`;
}

/** "09:30" from a `<input type="time">` to minutes past midnight. */
export function minutesFromInput(value: string): number | null {
  const m = /^(\d{1,2}):(\d{2})$/.exec(value.trim());
  if (!m) return null;
  const h = Number(m[1]);
  const mi = Number(m[2]);
  if (h > 23 || mi > 59) return null;
  return h * 60 + mi;
}

File: flutter_app/lib/core/time.dart
/// Every time and date in this app is shown in India Standard Time (UTC+05:30),
/// on a 12-hour clock, regardless of what offset the API sends or what time zone
/// the machine is set to. Internally everything is absolute UTC milliseconds.
library;

const int kMin = 60000;
const int kIstOffset = 330; // minutes east of UTC
const int kDayMs = 24 * 60 * kMin;

const _months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
const _days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
const _daysFull = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

final _iso = RegExp(r'^(\d{4})-(\d{2})-(\d{2})[T ](\d{2}):(\d{2})(?::(\d{2}))?(?:\.\d+)?(Z|[+-]\d{2}:?\d{2})?$');

int nowMs() => DateTime.now().millisecondsSinceEpoch;

/// Parse an ISO timestamp to UTC ms. A missing offset is read as IST.
int? parseISO(Object? value) {
  if (value == null) return null;
  final text = value.toString().trim();
  if (text.isEmpty) return null;
  final m = _iso.firstMatch(text);
  if (m == null) return DateTime.tryParse(text)?.millisecondsSinceEpoch;

  var offset = kIstOffset;
  final zone = m.group(7);
  if (zone != null) {
    if (zone == 'Z') {
      offset = 0;
    } else {
      final z = RegExp(r'([+-])(\d{2}):?(\d{2})').firstMatch(zone);
      if (z != null) {
        offset = (z.group(1) == '-' ? -1 : 1) * (int.parse(z.group(2)!) * 60 + int.parse(z.group(3)!));
      }
    }
  }
  return DateTime.utc(
        int.parse(m.group(1)!),
        int.parse(m.group(2)!),
        int.parse(m.group(3)!),
        int.parse(m.group(4)!),
        int.parse(m.group(5)!),
        int.parse(m.group(6) ?? '0'),
      ).millisecondsSinceEpoch -
      offset * kMin;
}

/// Shift into IST so the UTC accessors read as Indian wall-clock time.
DateTime ist(int t) => DateTime.fromMillisecondsSinceEpoch(t + kIstOffset * kMin, isUtc: true);

String p2(int n) => n < 10 ? '0$n' : '$n';
int _h12(DateTime d) => d.hour % 12 == 0 ? 12 : d.hour % 12;
String _mer(DateTime d) => d.hour >= 12 ? 'PM' : 'AM';
int _dow(DateTime d) => d.weekday % 7; // Sunday = 0, as the labels are ordered

/// `09:47 AM`
String clockTime(int? t) {
  if (t == null) return '--:--';
  final d = ist(t);
  return '${p2(_h12(d))}:${p2(d.minute)} ${_mer(d)}';
}

/// `9:47 AM` — no leading zero, for headline use.
String clockShort(int? t) {
  if (t == null) return '--:--';
  final d = ist(t);
  return '${_h12(d)}:${p2(d.minute)} ${_mer(d)}';
}

/// `03:00:07 PM`
String clockWithSeconds(int t) {
  final d = ist(t);
  return '${p2(_h12(d))}:${p2(d.minute)}:${p2(d.second)} ${_mer(d)}';
}

/// `17:30` — 24-hour, for the ETA tile.
String clock24(int t) {
  final d = ist(t);
  return '${p2(d.hour)}:${p2(d.minute)}';
}

/// `07/08/2026` — Indian day-first date format.
String dateDMY(int t) {
  final d = ist(t);
  return '${p2(d.day)}/${p2(d.month)}/${d.year}';
}

/// `Fri`
String shortDay(int t) => _days[_dow(ist(t))];

/// `FRIDAY` once upper-cased by the caller.
String weekday(int t) => _daysFull[_dow(ist(t))];

/// `06 Aug 2026`
String dateLabel(int t) {
  final d = ist(t);
  return '${p2(d.day)} ${_months[d.month - 1]} ${d.year}';
}

String dayKey(int t) {
  final d = ist(t);
  return '${d.year}-${p2(d.month)}-${p2(d.day)}';
}

/// `2026-08` — the key a stored day is filed and filtered under.
String monthKey(int t) {
  final d = ist(t);
  return '${d.year}-${p2(d.month)}';
}

/// `Aug 2026`
String monthLabel(int t) {
  final d = ist(t);
  return '${_months[d.month - 1]} ${d.year}';
}

/// `Aug 2026` from a stored `2026-08` key, for a month with no days left in it.
String monthLabelFromKey(String key) {
  final m = RegExp(r'^(\d{4})-(\d{2})$').firstMatch(key);
  if (m == null) return key;
  final index = int.parse(m.group(2)!) - 1;
  if (index < 0 || index > 11) return key;
  return '${_months[index]} ${m.group(1)}';
}

/// `07 Aug` — day and month only, for dense history rows.
String dayMonth(int t) {
  final d = ist(t);
  return '${p2(d.day)} ${_months[d.month - 1]}';
}

/// Midnight IST of the day containing `t`, as UTC ms.
int startOfDay(int t) {
  final d = ist(t);
  return DateTime.utc(d.year, d.month, d.day).millisecondsSinceEpoch - kIstOffset * kMin;
}

int startOfToday() => startOfDay(nowMs());

/// Midnight IST of a `YYYY-MM-DD` key.
int? startOfDayKey(String key) {
  final m = RegExp(r'^(\d{4})-(\d{2})-(\d{2})$').firstMatch(key.trim());
  if (m == null) return null;
  return DateTime.utc(int.parse(m.group(1)!), int.parse(m.group(2)!), int.parse(m.group(3)!)).millisecondsSinceEpoch -
      kIstOffset * kMin;
}

/// `7:32`
String hm(num ms) {
  final neg = ms < 0;
  final total = ms.abs() ~/ kMin;
  return '${neg ? '-' : ''}${total ~/ 60}:${p2(total % 60)}';
}

/// `7:32:09`
String hms(num ms) {
  final s = ms.abs() ~/ 1000;
  return '${s ~/ 3600}:${p2((s ~/ 60) % 60)}:${p2(s % 60)}';
}

/// Split for per-place animation: `["7", "32", "09"]`
List<String> hmsParts(num ms) {
  final s = ms.abs() ~/ 1000;
  return ['${s ~/ 3600}', p2((s ~/ 60) % 60), p2(s % 60)];
}

/// `8h 00m` — padded, for the goal headline.
String hoursMinutes(num ms) {
  final total = (ms.abs() / kMin).round();
  return '${total ~/ 60}h ${p2(total % 60)}m';
}

/// `1h 26m`
String human(num ms) {
  final total = (ms.abs() / kMin).round();
  final h = total ~/ 60;
  return '${h > 0 ? '${h}h ' : ''}${total % 60}m';
}

/// Current IST wall-clock as `"14:05"`.
String inputTimeNow() => clock24(nowMs());

/// "09:30" to minutes past midnight.
int? minutesFromInput(String value) {
  final m = RegExp(r'^(\d{1,2}):(\d{2})$').firstMatch(value.trim());
  if (m == null) return null;
  final h = int.parse(m.group(1)!);
  final mi = int.parse(m.group(2)!);
  if (h > 23 || mi > 59) return null;
  return h * 60 + mi;
}
